File: packages/react-native/ios/OpenOtaModule.swift
import ExpoModulesCore

public class OpenOtaModule: Module {

  public func definition() -> ModuleDefinition {
    Name("OpenOta")

    Events(
      OpenOta.downloadProgressEvent,
      OpenOta.updateStateEvent,
      OpenOta.previewRequestedEvent
    )

    Constants { OpenOta.constants() }

    OnCreate {
      OpenOta.emitter = { [weak self] name, body in self?.sendEvent(name, body) }
      OpenOta.startObservingLinks()
    }

    OnDestroy {
      OpenOta.emitter = nil
      OpenOta.stopObservingLinks()
    }

    Function("getStatus") { withNulls(OpenOta.status()) }

    Function("getPendingPreview") { OpenOta.pendingPreview() }

    Function("setChannel") { (channel: String?) in OpenOta.setChannel(channel) }

    Function("clearFailed") { OpenOta.clearFailed() }

    AsyncFunction("downloadUpdate") { (manifestJson: String, signatureBase64: String, url: String, promise: Promise) in
      OpenOta.downloadUpdate(
        manifestJson: manifestJson,
        signatureBase64: signatureBase64,
        url: url,
        onProgress: { [weak self] written, total in
          self?.sendEvent(OpenOta.downloadProgressEvent, ["written": written, "total": total])
        }
      ) { result in
        switch result {
        case .success(let value): promise.resolve(value)
        case .failure(let error): reject(promise, error)
        }
      }
    }

    AsyncFunction("applyUpdate") { (reload: Bool, promise: Promise) in
      do {
        try OpenOta.applyUpdate(reloadNow: reload)
        promise.resolve(nil)
      } catch {
        reject(promise, error)
      }
    }

    AsyncFunction("notifyAppReady") { (promise: Promise) in
      OpenOta.notifyAppReady()
      promise.resolve(nil)
    }

    AsyncFunction("reload") { (promise: Promise) in
      OpenOta.reload()
      promise.resolve(nil)
    }

    AsyncFunction("rollback") { (reason: String, promise: Promise) in
      OpenOta.rollback(reason: reason)
      promise.resolve(nil)
    }

    AsyncFunction("exitPreview") { (promise: Promise) in
      OpenOta.exitPreview()
      promise.resolve(nil)
    }

    AsyncFunction("takePendingEvents") { (promise: Promise) in
      promise.resolve(OpenOta.takePendingEvents())
    }

    AsyncFunction("handlePreviewLink") { (url: String, promise: Promise) in
      do {
        promise.resolve(try OpenOta.handleDeepLink(url))
      } catch {
        reject(promise, error)
      }
    }
  }
}

/// Every failure reaches JS as a typed code instead of an opaque exception.
private func reject(_ promise: Promise, _ error: Error) {
  if let error = error as? OtaError {
    promise.reject(error.code, error.message)
  } else {
    promise.reject("ERR_OTA_UNKNOWN", error.localizedDescription)
  }
}

/// JS expects explicit nulls where the SDK has no value.
private func withNulls(_ values: [String: Any?]) -> [String: Any] {
  values.mapValues { $0 ?? NSNull() }
}

File: packages/react-native/android/src/main/java/dev/openota/OpenOtaModule.kt
package dev.openota

import android.content.Context
import android.content.Intent
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class OpenOtaModule : Module() {

  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    Name("OpenOta")

    Events(
      OpenOta.EVENT_DOWNLOAD_PROGRESS,
      OpenOta.EVENT_UPDATE_STATE,
      OpenOta.EVENT_PREVIEW_REQUESTED,
    )

    Constants { OpenOta.constants(context) }

    OnCreate {
      OpenOta.attach(context)
      OpenOta.emitter = { name, body -> this@OpenOtaModule.sendEvent(name, body) }
      // A cold start opened by the deep link already consumed the intent.
      appContext.currentActivity?.intent?.let { runCatching { handleIntent(it) } }
    }

    OnDestroy { OpenOta.emitter = null }

    OnNewIntent { intent -> runCatching { handleIntent(intent) } }

    Function("getStatus") { OpenOta.status() }

    Function("getPendingPreview") { OpenOta.pendingPreview() }

    Function("setChannel") { channel: String? -> OpenOta.setChannel(channel) }

    Function("clearFailed") { OpenOta.clearFailed() }

    AsyncFunction("downloadUpdate") { manifestJson: String, signatureBase64: String, url: String, promise: Promise ->
      guard(promise) { OpenOta.downloadUpdate(manifestJson, signatureBase64, url) }
    }

    AsyncFunction("applyUpdate") { reload: Boolean, promise: Promise ->
      guard(promise) { OpenOta.applyUpdate(reload); null }
    }

    AsyncFunction("notifyAppReady") { promise: Promise ->
      guard(promise) { OpenOta.notifyAppReady(); null }
    }

    AsyncFunction("reload") { promise: Promise ->
      guard(promise) { OpenOta.reload(); null }
    }

    AsyncFunction("rollback") { reason: String, promise: Promise ->
      guard(promise) { OpenOta.rollback(reason); null }
    }

    AsyncFunction("exitPreview") { promise: Promise ->
      guard(promise) { OpenOta.exitPreview(); null }
    }

    AsyncFunction("takePendingEvents") { promise: Promise ->
      guard(promise) { OpenOta.takePendingEvents() }
    }

    AsyncFunction("handlePreviewLink") { url: String, promise: Promise ->
      guard(promise) { OpenOta.handleDeepLink(url) }
    }
  }

  private fun handleIntent(intent: Intent) {
    if (intent.action == Intent.ACTION_VIEW) OpenOta.handleDeepLink(intent.data?.toString())
  }

  /** Every failure reaches JS as a typed code instead of an opaque exception. */
  private inline fun guard(promise: Promise, body: () -> Any?) {
    try {
      promise.resolve(body())
    } catch (e: OtaError) {
      promise.reject(e.code, e.message, e)
    } catch (t: Throwable) {
      promise.reject("ERR_OTA_UNKNOWN", t.message ?: t.javaClass.simpleName, t)
    }
  }
}
